devito/interpolators.py

```diff
@@ -0,0 +1,354 @@
+import sympy
+import numpy as np
+
+from abc import ABC, abstractmethod
+
+from devito.equation import Eq, Inc
+from devito.logger import warning
+from devito.symbolics import retrieve_function_carriers, indexify, INT
+from devito.tools import Evaluable, powerset, flatten, prod
+from devito.types.basic import Scalar
+from devito.types.dense import SubFunction
+from devito.types.dimension import ConditionalDimension, Dimension, DefaultDimension
+
+
+class UnevaluatedSparseOperation(Evaluable):
+
+    """
+    Represents an Injection or an Interpolation operation performed on a
+    SparseFunction. Evaluates to a list of Eq objects.
+    objects.
+
+    Parameters
+    ----------
+    interpolator : Interpolator
+        Interpolator object that will be used to evaluate the Operation
+    *args, **kwargs
+        The arguments passed to the corresponding method
+    """
+    subdomain = None
+
+    def __init__(self, interpolator, *args, **kwargs):
+        self.interpolator = interpolator
+        self._args = args
+        self._kwargs = kwargs
```

Comment:
why do you need to store args and kwargs?

Reply:
If anything was passed to the inject method on the sparse function, it needs to be passed to the interpolator. So it has to be stored.

Reply:
If there is explicit arguments/keyword needed should be explicit, if not this is not needed, specially without doctoring saying why should be in args/kwargs.

Reply:
This bit of code doesn't need to know what those arguments are. So why should it be explicit?This is just forwarding the arguments to where they are actually used. Aren't you making the opposite point of what @FabioLuporini is making here : https://github.com/devitocodes/devito/pull/1025#discussion_r363207849 ?

Reply:
No I don't think it's opposite or related to his comment. I am talking about readability and maintainability, there is currently no way to know what `init` and therefor Interp/Inject expect as input, only some anonymous args/kwargs, so it need  to either be explicit or add a complete doctoring that says what it expects as an input.

Reply:
I ask you to look at the complete change I made in the previous commit in entirety. This object is created in `Interpolator.inject` and serves to store state to call `Interpolator._inject`. The docstring is present where this object is created and is right next to the code that consumes those arguments. I can add it here too if you want, but I think it is unnecessary. 
I would like to move as much of the logic to the interpolator class as possible so that these classes can get much closer to the `Future` class that they conceptually are. Any ideas for reducing code in the UnevaluatedSparseOperation/Injection/Interpolation classes are welcome.

Reply:
`args` and `kwargs` are not used anywhere within the class and just dumping everything here in case some subclass need it is not good even for Future

Reply:
>The docstring is present where this object is created a
That's not ok. The doctoring for a class goes with the class, can't expect people to dig to figure out where someone creates an instance of it to figure out  what the input is.

Reply:
> args and kwargs are not used anywhere within the class and just dumping everything here in case some subclass need it is not good even for Future
At first look, this is a valid concern. However, I've been trying to figure out a way to move `evaluate` to this class - which would make the subclasses much simpler. The question is of how the subclass will modify the behaviour between inject and interpolate. I don't want to write the name of the method as a string in the subclass and I can't think of another way to do it. I'm still hoping to find a way to do it but I'll write this as another TODO in the issue. 

For the medium to long term future, I envision that this entire series of classes could be abstracted out into decorator-based setups - a decorator feels like the right abstraction for this situation, but it would be too much work to go into for this PR. 

> That's not ok. The doctoring for a class goes with the class, can't expect people to dig to figure out where someone creates an instance of it to figure out what the input is.

In my opinion, these classes are not intended for direct use by the end user. The inject/interpolate functions on the Interpolator are. Either both functions should have the same doctstring, or the user-facing functions should have a doctsring and the internal classes could reference that. Do you want me to add a docstring here that says "See LinearInterpolator.inject"?

Reply:
>In my opinion, these classes are not intended for direct use by the end user. 

That's not the point, most of the API is documented wether it's internal or user facing so that people can contribute and maintain it properly and not have to dig everywhere and goes what something is doing. This is a class with a constructor, it needs a doctoring, and I don't like pointing to "just look at how we call it there that works but can maybe do otherwise".